### PR TITLE
Treat default weblinks same as any other weblinks

### LINKED ISSUE #3354: Delete default weblinks in catalog
Be able to delete default weblinks in catalog in Lens 5

### PR DESCRIPTION
**Pull request by Nokel81**
Signed-off-by: Sebastian Malton <sebastian@malton.name>

fixes #3354

### CODE REVIEW

File: src/main/catalog-sources/weblinks.ts [hunk 1]
```diff
@@ -19,84 +19,75 @@
...
-      status: {
-        phase: "available",
-        active: true
-      }
-    })));
-    weblinks.push(...defaultLinks);
+    const seenWeblinks = new Set<string>();
+
+    for (const weblinkData of links) {
+      seenWeblinks.add(weblinkData.id);
```
Comment (jim-docker) [inline src/main/catalog-sources/weblinks.ts:70]: looks like you're not actually doing anything with this

File: src/migrations/weblinks-store/5.1.1-beta.1.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import { docsUrl, slackUrl } from "../../common/vars";
+import type { WeblinkData } from "../../common/weblink-store";
+import type { MigrationDeclaration } from "../helpers";
+
+export default {
+  version: "5.1.1-beta.1",
+  run(store) {
+    const weblinks: WeblinkData[] = store.get("weblinks");
+
+    weblinks.push(
```
Comment (jim-docker) [inline src/migrations/weblinks-store/5.1.1-beta.1.ts]: weblinks can be undefined

**Review by jim-docker — COMMENTED** [review]
I get this for `make dev`:
```
[0] Running 5.1.1-beta.1 migration for /Users/jimehrismann/Library/Application Support/OpenLensDev/lens-cluster-store.json
[0] info: [STORE]: LOADED from /Users/jimehrismann/Library/Application Support/OpenLensDev/lens-cluster-store.json
[0] info: [STORE]: LOADED from /Users/jimehrismann/Library/Application Support/OpenLensDev/lens-hotbar-store.json
[0] info: [STORE]: LOADED from /Users/jimehrismann/Library/Application Support/OpenLensDev/lens-extensions.json
[0] info: [STORE]: LOADED from /Users/jimehrismann/Library/Application Support/OpenLensDev/lens-filesystem-provisioner-store.json
[0] Running 5.1.1-beta.1 migration for /Users/jimehrismann/Library/Application Support/OpenLensDev/lens-weblink-store.json
[0] (node:9054) UnhandledPromiseRejectionWarning: Error: Something went wrong during the migration! Changes applied to the store until this failed migration will be restored. TypeError: Cannot read property 'push' of undefined
[0]     at Conf._migrate (/Users/jimehrismann/git/lensapp/lens/node_modules/conf/dist/source/index.js:407:23)
[0]     at new Conf (/Users/jimehrismann/git/lensapp/lens/node_modules/conf/dist/source/index.js:144:18)
[0]     at WeblinkStore.load (webpack-internal:///./src/common/base-store.ts:74:28)
[0]     at new WeblinkStore (webpack-internal:///./src/common/weblink-store.ts:62:14)
[0]     at Function.createInstance (webpack-internal:///./src/common/utils/singleton.ts:48:47)
[0]     at App.eval (webpack-internal:///./src/main/index.ts:175:73)
[0]     at processTicksAndRejections (internal/process/task_queues.js:94:5)
[0] (node:9054) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 8)
[0] (node:9054) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```
It was "migrating" from master (I pulled master and ran `make dev` before pulling this branch)

**Comment by Nokel81** [comment]
That is probably the "weblinks" was undefined.

**Comment by jim-docker** [comment]
> That is probably the "weblinks" was undefined.

yes, but shouldn't there have been the default weblinks in the store?

**Comment by Nokel81** [comment]
@jim-docker That is what this PR is trying to implement. There weren't any default weblinks in the store.

**Comment by jim-docker** [comment]
> @jim-docker That is what this PR is trying to implement. There weren't any default weblinks in the store.

So you didn't test this?

**Comment by Nokel81** [comment]
I did, but I guess I didn't test it enough.

**Comment by Nokel81** [comment]
Namely, if I already had ever added a weblink that issue doesn't appear.

File: src/main/catalog-sources/weblinks.ts [hunk 3]
```diff
@@ -19,84 +19,83 @@
...
   }
 }
 
 
 export function syncWeblinks() {
   const weblinkStore = WeblinkStore.getInstance();
-  const weblinks = observable.array(defaultLinks);
+  const webLinkEntities = observable.map<string, [WebLink, Disposer]>();
+
+  function periodicallyCheckLink(link: WebLink): Disposer {
```
Comment (jim-docker) [inline src/main/catalog-sources/weblinks.ts:52]: This looks scary to me. A timer for each weblink? An extension could programmatically generate a lot of links? I think this is outside the scope of the issue anyway

File: src/migrations/weblinks-store/5.1.1-beta.1.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+import { docsUrl, slackUrl } from "../../common/vars";
+import type { WeblinkData } from "../../common/weblink-store";
+import type { MigrationDeclaration } from "../helpers";
+
+export default {
+  version: "5.1.1-beta.1",
+  run(store) {
+    const weblinksRaw: any = store.get("weblinks");
+    const weblinks = (Array.isArray(weblinksRaw) ? weblinksRaw : []) as WeblinkData[];
```
Comment (jim-docker) [inline src/migrations/weblinks-store/5.1.1-beta.1.ts]: seems like it could be more concise?

Comment (Nokel81) [reply src/migrations/weblinks-store/5.1.1-beta.1.ts]: It could be. All depends on how data paranoid we want to be. In theory I could even do some validation of the contents of that array. But I felt that throwing away items that were clearly wrong (like strings) was good enough.

File: src/main/catalog-sources/weblinks.ts (hunk 3 above)
Comment (Nokel81) [reply src/main/catalog-sources/weblinks.ts:52]: By "outside the scope" do you mean you think I should revert this and only check each link once?

Comment (jim-docker) [reply src/main/catalog-sources/weblinks.ts:52]: I guess you want `setInterval`?
I think I would revert it, seems more of a feature. They are running all the time? Or just when the catalog is displayed?

Comment (Nokel81) [reply src/main/catalog-sources/weblinks.ts:52]: I did mean `setInterval` yes. They would be running all the time, I guess I could make it so that there is an initial spread so that there is some smoothness to the checking (and not spikes every 10 minutes).

I would still say it is a bug because the catalog displays the phase and it should show as "unavailable" if it is actually unavailable.

Comment (jim-docker) [reply src/main/catalog-sources/weblinks.ts:52]: How often do weblinks become unavailable? Maybe I'm missing something but I just don't see the value in using cycles to track this.

Comment (Nokel81) [reply src/main/catalog-sources/weblinks.ts:52]: Anytime the website goes down. it would become unavailable.

Comment (Nokel81) [reply src/main/catalog-sources/weblinks.ts:52]: @jakolehm Should we check them periodically or only once per set change (which is what it was before).

Comment (Nokel81) [reply src/main/catalog-sources/weblinks.ts:52]: Jari says that once initially and then every hour or so.

**Review by jim-docker — APPROVED** [review]
The extraneous hotbar field in lens-weblink-store.json should probably be cleaned up in another PR?

**Comment by Nokel81** [comment]
@jim-docker Yes I think it should, probably by fixing `BaseStore` itself.
